File: app/src/main/java/com/example/jianming/views/YImageSlider.java
package com.example.jianming.views;

import android.content.Context;
import android.graphics.Color;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import com.example.jianming.Utils.DIOptionsNoneScaled;
import com.example.jianming.myapplication.R;
import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.download.ImageDownloader;

/**
 * Created by Jianming on 2015/9/22.
 */
public class YImageSlider extends ViewGroup implements YImageView.EdgeListener {
    public YImageSlider(Context context) {
        super(context);
        init(context);
    }

    public final static int SPLITE_W = 48;

    public YImageSlider(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public YImageSlider(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }


    private void init(Context context) {
        contentView = new YImageView(context, this, 0);
        contentView.setEdgeListener(this);
        hideLeft = new YImageView(context, this, -1);
        hideRight = new YImageView(context, this, 1);

        hideLeft.setEdgeListener(this);
        hideRight.setEdgeListener(this);

        backButton = new ImageView(context);
        nextButton = new ImageView(context);


        backButton.setImageResource(R.drawable.ic_keyboard_arrow_left_black_48dp);
        nextButton.setImageResource(R.drawable.ic_keyboard_arrow_right_black_48dp);
        backButton.setBackgroundColor(Color.parseColor("#80000000"));
        nextButton.setBackgroundColor(Color.parseColor("#80000000"));

        nextButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                contentView.doNextImgAnim();
            }
        });

        backButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                contentView.doBackImgAnim();
            }
        });

        addView(contentView);
        addView(hideLeft);
        addView(hideRight);

        addView(backButton);
        addView(nextButton);

        line31 = new View(context);
        line32 = new View(context);

        line31.setBackgroundColor(Color.parseColor("green"));
        line32.setBackgroundColor(Color.parseColor("blue"));

//        addView(line31);
//        addView(line32);
//        ImageLoader.getInstance().displayImage(ImageDownloader.Scheme.DRAWABLE.wrap(pics[index] + ""), hideLeft, DIOptionsNoneScaled.getInstance().getOptions());
//        ImageLoader.getInstance().displayImage(ImageDownloader.Scheme.DRAWABLE.wrap(pics[index + 1] + ""), contentView, DIOptionsNoneScaled.getInstance().getOptions());
//        ImageLoader.getInstance().displayImage(ImageDownloader.Scheme.DRAWABLE.wrap(pics[index + 2] + ""), hideRight, DIOptionsNoneScaled.getInstance().getOptions());
    }

    private YImageView contentView, hideLeft, hideRight;

    public YImageView getHideLeft() {
        return hideLeft;
    }

    public YImageView getHideRight() {
        return hideRight;
    }

    private ImageView backButton;

    private ImageView nextButton;

    private View line31, line32;

    public YImageView getContentView() {
        return contentView;
    }

    int index = 0;
    int pics[] = {R.drawable.su27_1, R.drawable.su27_2, R.drawable.su27_3, R.drawable.su27_4,
            R.drawable.su27_1, R.drawable.su27_2, R.drawable.su27_3, R.drawable.su27_4};

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        int width = r - l;
        int height = b - t;
        contentView.layout(0, 0, width, height);
        hideLeft.layout(0, 0, width, height);
        hideRight.layout(0, 0, width, height);

        backButton.layout(0, 0, 48, 48);
        nextButton.layout(width - 48, 0, width, 48);

//        line31.layout(width / 3, 0, width / 3 + 1, height);
//        line32.layout(width * 2 / 3, 0, width * 2 / 3 + 1, height);
    }

    @Override
    public void onXEdge(YImageView yImageView) {

    }

    @Override
    public void onYEdge(YImageView yImageView) {

    }

    public void setHideLeftSrc(int index) {
        if (imgChangeListener != null) {
            String src = imgChangeListener.getImgSrcByIndex(index - 1, this);
            YImageView yImageView = getHideLeft();
            if (src != null) {
                ImageLoader.getInstance().displayImage(src, yImageView, DIOptionsNoneScaled.getInstance().getOptions());
                yImageView.setDisplay();
            } else {
                yImageView.setNoDisplay();
            }
        }
    }

    public void setContentSrc(int index) {

        if (imgChangeListener != null) {
            String src = imgChangeListener.getImgSrcByIndex(index, this);
            YImageView yImageView = getContentView();
            if (src != null) {
                ImageLoader.getInstance().displayImage(src, yImageView, DIOptionsNoneScaled.getInstance().getOptions());
                yImageView.setDisplay();
            } else {
                yImageView.setNoDisplay();
            }
        }
    }

    public void setHideRightSrc(int index) {
        if (imgChangeListener != null) {
            String src = imgChangeListener.getImgSrcByIndex(index + 1, this);
            YImageView yImageView = getHideRight();
            if (src != null) {
                ImageLoader.getInstance().displayImage(src, yImageView, DIOptionsNoneScaled.getInstance().getOptions());
                yImageView.setDisplay();
            } else {
                yImageView.setNoDisplay();
            }
        }
    }

    public interface ImgChangeListener {
        String onGetBackImg(YImageSlider yImageSlider);

        String onGetNextImg(YImageSlider yImageSlider);

        String getImgSrcByIndex(int index, YImageSlider yImageSlider);
    }

    public void setImgChangeListener(ImgChangeListener imgChangeListener) {
        this.imgChangeListener = imgChangeListener;
    }

    ImgChangeListener imgChangeListener = null;

    @Override
    public void onGetBackImg(YImageView yImageView) {
        YImageView tmp = contentView;
        contentView = hideLeft;
        hideLeft = hideRight;
        hideRight = tmp;

        contentView.setLocationIndex(0);
        hideLeft.setLocationIndex(-1);
        hideRight.setLocationIndex(1);
        alingLeftOrRight = 0;
        String imgUrl;

        if (imgChangeListener != null) {
            imgUrl = imgChangeListener.onGetBackImg(this);
        } else {
            index--;
            if (index < 0) {
                index = 3;
            }
            imgUrl = ImageDownloader.Scheme.DRAWABLE.wrap(pics[index] + "");
        }
        if (imgUrl != null) {
            ImageLoader.getInstance().displayImage(imgUrl, getHideLeft(), DIOptionsNoneScaled.getInstance().getOptions());
            getHideLeft().setDisplay();
        } else {
            getHideLeft().setNoDisplay();
        }

    }

    @Override
    public void onGetNextImg(YImageView yImageView) {
        YImageView tmp = contentView;
        contentView = hideRight;
        hideRight = hideLeft;
        hideLeft = tmp;

        contentView.setLocationIndex(0);
        hideLeft.setLocationIndex(-1);
        hideRight.setLocationIndex(1);
        alingLeftOrRight = 1;
        String imgUrl;
        if (imgChangeListener != null) {
            imgUrl = imgChangeListener.onGetNextImg(this);
        } else {
            index++;
            if (index > 2) {
                index = 0;
            }
            imgUrl = ImageDownloader.Scheme.DRAWABLE.wrap(pics[index + 2] + "");
        }
        if (imgUrl != null) {
            ImageLoader.getInstance().displayImage(imgUrl, hideRight, DIOptionsNoneScaled.getInstance().getOptions());
            hideRight.setDisplay();
        } else {
            hideRight.setNoDisplay();
        }
    }

    public int getAlingLeftOrRight() {
        return alingLeftOrRight;
    }

    private int alingLeftOrRight = 0;


}

File: app/src/main/java/org/nanking/knightingal/kslideviewlib/YImageSlider.kt
package org.nanking.knightingal.kslideviewlib

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Color
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import org.knightingal.flow1000.client.R


class YImageSlider : ViewGroup {
    companion object {

        val TAG = "YImageSlider"
        val SPLITE_W = 48
    }
    constructor(context:Context) : super(context) {
        init(context)
    }


    constructor(context:Context, attrs:AttributeSet):super(context, attrs) {
        init(context)
    }

    constructor(context:Context, attrs:AttributeSet, defStyleAttr:Int):super(context, attrs, defStyleAttr) {
        init(context);
    }


    private fun init(context:Context) {
        contentView = YImageView(context, this, 0, )
        hideLeft = YImageView(context, this, -1, )
        hideRight = YImageView(context, this, 1, )

        contentView.postGetBackImg = this::onGetBackImg
        hideLeft.postGetBackImg = this::onGetBackImg
        hideRight.postGetBackImg = this::onGetBackImg

        contentView.postGetNextImg = { onGetNextImg() }
        hideLeft.postGetNextImg = { onGetNextImg() }
        hideRight.postGetNextImg = { onGetNextImg() }

        contentView.postXEdgeEvent = this.onXEdge
        hideLeft.postXEdgeEvent = this.onXEdge
        hideRight.postXEdgeEvent = this.onXEdge

        contentView.postYEdgeEvent = { onYEdge() }
        hideLeft.postYEdgeEvent = { onYEdge() }
        hideRight.postYEdgeEvent = { onYEdge() }

        backButton = ImageView(context)
        nextButton = ImageView(context)


        backButton.setImageResource(R.drawable.ic_keyboard_arrow_left_black_48dp)
        nextButton.setImageResource(R.drawable.ic_keyboard_arrow_right_black_48dp)
        backButton.setBackgroundColor(Color.parseColor("#80000000"))
        nextButton.setBackgroundColor(Color.parseColor("#80000000"))

        backButton.visibility = View.INVISIBLE

        nextButton.setOnClickListener { contentView.doNextImgAnim() }

        backButton.setOnClickListener { contentView.doBackImgAnim() }

        addView(contentView)
        addView(hideLeft)
        addView(hideRight)

        addView(backButton)
        addView(nextButton)
    }

    lateinit var contentView:YImageView
    lateinit var hideLeft:YImageView
    lateinit var hideRight:YImageView


    private lateinit var backButton:ImageView

    private lateinit var nextButton:ImageView



    override fun onLayout(changed:Boolean, l:Int, t:Int, r:Int, b:Int) {
        val width = r - l
        val height = b - t
        contentView.layout(0, 0, width, height)
        hideLeft.layout(0, 0, width, height)
        hideRight.layout(0, 0, width, height)

        backButton.layout(0, height / 2 - 24, 48, height / 2 + 24)
        nextButton.layout(width - 48, height / 2 - 24, width, height / 2 + 24)
    }

    val onXEdge : ()->Unit = {
        Log.d(TAG, "onXEdge")
    }

    fun onYEdge() {
        Log.d(TAG, "onYEdge")
    }

    fun setHideLeftSrc(index:Int) {
        if (imgChangeListener != null) {
            val src = imgChangeListener?.getImgSrcByIndex(index - 1, this)
            val yImageView = hideLeft
            if (src != null) {
                val mb = BitmapFactory.decodeFile(src)
                yImageView.setImageBitmap(mb)
                yImageView.isDisplay = true
            } else {
                yImageView.isDisplay = false
            }
        }
    }

    fun setContentSrc(index:Int) {

        if (imgChangeListener != null) {
            val src = imgChangeListener?.getImgSrcByIndex(index, this);
            val yImageView = contentView
            if (src != null) {
                val mb = BitmapFactory.decodeFile(src)
                yImageView.setImageBitmap(mb)
                yImageView.isDisplay = true
            } else {
                yImageView.isDisplay = false
            }
        }
    }

    fun setHideRightSrc(index:Int) {
        if (imgChangeListener != null) {
            val src = imgChangeListener?.getImgSrcByIndex(index + 1, this)
            val yImageView = hideRight
            if (src != null) {
                val mb = BitmapFactory.decodeFile(src)
                yImageView.setImageBitmap(mb)
//                ImageLoader.getInstance().displayImage(src, yImageView, DIOptionsNoneScaled.options)
                yImageView.isDisplay = true
            } else {
                yImageView.isDisplay = false
            }
        }
    }

    public interface ImgChangeListener {
        fun onGetBackImg(yImageSlider:YImageSlider):String

        fun onGetNextImg(yImageSlider:YImageSlider):String

        fun getImgSrcByIndex(index:Int, yImageSlider:YImageSlider):String
    }


    var imgChangeListener:ImgChangeListener? = null

    fun onGetBackImg() {
        Log.d(TAG, "onGetBackImg")
        var tmp = contentView
        contentView = hideLeft
        hideLeft = hideRight
        hideRight = tmp

        contentView.locationIndex = (0)
        hideLeft.locationIndex = (-1)
        hideRight.locationIndex = (1)
        alingLeftOrRight = 0
        var imgUrl:String? = null

        if (imgChangeListener != null) {
            imgUrl = imgChangeListener?.onGetBackImg(this)
        }
        if (imgUrl != null) {
            if (nextButton.visibility == View.INVISIBLE) {
                nextButton.visibility = (View.VISIBLE)
            }
            val mb = BitmapFactory.decodeFile(imgUrl)
            hideLeft.setImageBitmap(mb)

//            ImageLoader.getInstance().displayImage(imgUrl, hideLeft, DIOptionsNoneScaled.options)
            hideLeft.isDisplay=(true)
        } else {
            backButton.visibility = (View.INVISIBLE)
//            ImageLoader.getInstance().displayImage(null, hideLeft)
            hideLeft.isDisplay = (false);
        }
        requestLayout()

    }

    fun onGetNextImg() {
        Log.d(TAG, "onGetNextImg");
        var tmp = contentView
        contentView = hideRight
        hideRight = hideLeft
        hideLeft = tmp

        contentView.locationIndex = (0)
        hideLeft.locationIndex = (-1)
        hideRight.locationIndex = (1)
        alingLeftOrRight = 1
        var imgUrl:String? = null
        if (imgChangeListener != null) {
            imgUrl = imgChangeListener?.onGetNextImg(this);
        }
        if (imgUrl != null) {
            if (backButton.visibility == View.INVISIBLE) {
                backButton.visibility = (View.VISIBLE)
            }
            val mb = BitmapFactory.decodeFile(imgUrl)
            hideRight.setImageBitmap(mb)
            hideRight.isDisplay = (true)
        } else {
            nextButton.visibility = (View.INVISIBLE)
            hideRight.isDisplay = (false)
        }
        requestLayout()
    }


    var alingLeftOrRight:Int = 0;


}
